presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/WindowFilterPushDown.java

```diff
@@ -87,150 +92,189 @@ private Rewriter(PlanNodeIdAllocator idAllocator, Metadata metadata, Session ses
+                    Optional<PlanNode> filterNode = transformFilterNode(node, source, rowNumberSymbol, upperBound.get());
+                    return filterNode.orElse(source);
                 }
             }
+            else if (source instanceof WindowNode && canOptimizeWindowFunction((WindowNode) source)) {
+                WindowNode windowNode = (WindowNode) source;
+                Symbol rowNumberSymbol = getOnlyElement(windowNode.getWindowFunctions().entrySet()).getKey();
+                Optional<Integer> upperBound = extractUpperBound(tupleDomain, rowNumberSymbol);
 
-            if (upperBound.isPresent() && upperBound.get() <= Integer.MAX_VALUE) {
-                return Optional.of(upperBound.get().intValue());
+                if (upperBound.isPresent()) {
+                    source = convertToTopNRowNumber(windowNode, upperBound.get());
+                    Optional<PlanNode> filterNode = transformFilterNode(node, source, rowNumberSymbol, upperBound.get());
+                    return filterNode.orElse(source);
+                }
             }
-            return Optional.empty();
+
+            return context.replaceChildren(node, ImmutableList.of(source));
         }
 
-        private static boolean canOptimizeWindowFunction(WindowNode node)
+        private Optional<PlanNode> transformFilterNode(FilterNode filterNode, PlanNode source, Symbol rowNumberSymbol, int upperBound)
         {
-            if (node.getWindowFunctions().size() != 1) {
-                return false;
+            ExtractionResult extractionResult = fromPredicate(metadata, session, filterNode.getPredicate(), types);
+            TupleDomain<Symbol> tupleDomain = extractionResult.getTupleDomain();
+
+            if (!isEqualRange(tupleDomain, rowNumberSymbol, upperBound)) {
+                return Optional.of(new FilterNode(filterNode.getId(), source, filterNode.getPredicate()));
             }
-            Symbol rowNumberSymbol = getOnlyElement(node.getWindowFunctions().entrySet()).getKey();
-            return isRowNumberSignature(node.getSignatures().get(rowNumberSymbol));
+
+            // Remove the row number domain because it is absorbed into the node
+            Map<Symbol, Domain> newDomains = tupleDomain.getDomains().entrySet().stream()
+                    .filter(entry -> !entry.getKey().equals(rowNumberSymbol))
+                    .collect(toMap(Map.Entry::getKey, Map.Entry::getValue));
+
+            // Construct a new predicate
+            TupleDomain<Symbol> newTupleDomain = TupleDomain.withColumnDomains(newDomains);
+            Expression newPredicate = ExpressionUtils.combineConjuncts(
+                    extractionResult.getRemainingExpression(),
+                    toPredicate(newTupleDomain, types));
+
+            if (newPredicate.equals(BooleanLiteral.TRUE_LITERAL)) {
+                return Optional.empty();
+            }
+            return Optional.of(new FilterNode(idAllocator.getNextId(), source, newPredicate));
```

Comment:
idAllocator.getNextId() => filterNode.getId() to be consistent with the above return site, or vice versa